src/joybus.c

```diff
@@ -57,293 +71,224 @@ static volatile struct SI_regs_s * const SI_regs = (struct SI_regs_s *)0xa480000
-    memcpy(UncachedAddr(input_aligned), input, JOYBUS_BLOCK_SIZE);
+static void joybus_msg_send(joybus_msg_t *msg) {
+    assert((SI_regs->status & (SI_STATUS_DMA_BUSY | SI_STATUS_IO_BUSY)) == 0);
 
-    /* Be sure another thread doesn't get into a resource fight */
-    disable_interrupts();
-
-    __SI_DMA_wait();
-
-    SI_regs->DRAM_addr = input_aligned; // only cares about 23:0
+    data_cache_hit_writeback(msg->input, JOYBUS_BLOCK_SIZE);
+    SI_regs->DRAM_addr = msg->input;
     MEMORY_BARRIER();
     SI_regs->PIF_addr_write = PIF_RAM;
     MEMORY_BARRIER();
-
-    __SI_DMA_wait();
-
-    data_cache_hit_writeback_invalidate(output_aligned, JOYBUS_BLOCK_SIZE);
-
-    SI_regs->DRAM_addr = output_aligned;
-    MEMORY_BARRIER();
-    SI_regs->PIF_addr_read = PIF_RAM;
-    MEMORY_BARRIER();
-
-    __SI_DMA_wait();
-
-    /* Now that we've copied, its safe to let other threads go */
-    enable_interrupts();
-
-    memcpy(output, UncachedAddr(output_aligned), JOYBUS_BLOCK_SIZE);
+    joybus_state = JOYBUS_STATE_SENDING;
 }
 
 /**
- * @brief Read the status of the EEPROM.
- *
- * The EEPROM status response contains a byte-swapped identifier half-word that
- * indicates which size EEPROM is present on the cartridge and a status byte.
+ * @brief Receive a joybus reply from the PIF
+ * 
+ * @note This function must be called with interrupts disabled and SI must be idle
  *
- * @return the normalized Joybus EEPROM status response data.
+ * @param      msg    Message for which a reply is pending
  */
-static uint32_t eeprom_status( void )
-{
-    const uint64_t input[JOYBUS_BLOCK_DWORDS] =
-    {
-        0x00000000ff010300,
-        0xfffffffffe000000,
-        0,
-        0,
-        0,
-        0,
-        0,
-        1
-    };
-    uint64_t output[JOYBUS_BLOCK_DWORDS];
-
-    joybus_exec( input, output );
-
-    uint8_t * recv_bytes = (uint8_t *)&output[1];
-
-    return (
-        /* Intentional un-swap of identifier bytes */
-        ((uint32_t)recv_bytes[1] << 16) |
-        ((uint32_t)recv_bytes[0] << 8)  |
-        recv_bytes[2]
-    );
+static void joybus_msg_recv(joybus_msg_t *msg) {
+    assert((SI_regs->status & (SI_STATUS_DMA_BUSY | SI_STATUS_IO_BUSY)) == 0);
+
+    // Start a DMA transfer into the global temporary buffer. We just need
+    // one buffer for all messages as there can be only one ongoing joybus
+    // message at a time (it is a master/slave protocol).
+    data_cache_hit_invalidate(joybus_outbuf, JOYBUS_BLOCK_SIZE);
+    SI_regs->DRAM_addr = joybus_outbuf;
+    MEMORY_BARRIER();
+    SI_regs->PIF_addr_read = PIF_RAM;
+    MEMORY_BARRIER();
+    joybus_state = JOYBUS_STATE_RECEIVING;
 }
 
 /**
- * @brief Probe the EEPROM interface on the cartridge.
- *
- * Inspect the identifier half-word of the EEPROM status response to
- * determine which EEPROM save type is available (if any).
- *
- * @return which EEPROM type was detected on the cartridge.
+ * @brief Check where there are new messages to send
```

Comment:
It might be nice to add a comment to signify interrupts must be disabled when calling `joybus_state` manipulating functions.

Reply:
Done